pkg/image/controller/metrics.go

```diff
@@ -0,0 +1,192 @@
+	} else {
+		c.importErrorCounts[*defaultErrorInfoReason(errInfo, err)]++
+	}
+}
+
+// Collect is supposed to be called by the metrics callector. It returns the actual state of counters.
+func (c *ImportMetricCounter) Collect() (metrics.ImportSuccessCounts, metrics.ImportErrorCounts, error) {
+	c.counterMutex.Lock()
+	defer c.counterMutex.Unlock()
+
+	success := metrics.ImportSuccessCounts{}
+	for registry, count := range c.importSuccessCounts {
+		success[registry] = count
+	}
+
+	failures := metrics.ImportErrorCounts{}
+	for info, count := range c.importErrorCounts {
+		failures[info] = count
+	}
+
+	return success, failures, nil
+}
+
+// getIsImportRepositoryInfo returns an import error info if the given isi contains repository request.
+// If the request succeeded, the its Reason will be empty.
+func getIsImportRepositoryInfo(isi *imageapi.ImageStreamImport) *metrics.ImportErrorInfo {
+	if isi.Status.Repository == nil || isi.Spec.Repository == nil {
+		return nil
+	}
+	ref := isi.Spec.Repository.From
+	if ref.Kind != "DockerImage" {
+		return nil
+	}
+	imgRef, err := imageapi.ParseDockerImageReference(ref.Name)
+	if err != nil {
+		utilruntime.HandleError(fmt.Errorf(
+			"failed to parse isi.spec.repository.objectreference.from.name %q: %v",
+			ref.Name, err))
+		return nil
+	}
+
+	info := mkErrorInfo(imgRef.DockerClientDefaults().Registry, &isi.Status.Repository.Status)
+	return &info
+}
+
+// enumerateIsImportStatuses iterates over images of the given image stream import. For any valid recorded
+// import the cb callback will be called with the obtains information.
+// If the image import is successful, the object passed to the cb will contain empty Reason.
+func enumerateIsImportStatuses(isi *imageapi.ImageStreamImport, cb func(*metrics.ImportErrorInfo)) {
+	if len(isi.Status.Images) == 0 {
+		return
+	}
+
+	for i, status := range isi.Status.Images {
+		var (
+			imgRef imageapi.DockerImageReference
+			err    error
+		)
+
+		// TODO shall we always prefer docker image reference from status?
+		if status.Image == nil {
+			if i >= len(isi.Spec.Images) {
+				continue
+			}
+			imgSpec := &isi.Spec.Images[i]
+			if imgSpec.From.Kind != "DockerImage" {
+				continue
+			}
+			imgRef, err = imageapi.ParseDockerImageReference(imgSpec.From.Name)
+			if err != nil {
+				utilruntime.HandleError(fmt.Errorf(
+					"failed to parse isi.spec.images[%d].from.name %q: %v", i, imgSpec.From.Name, err))
+				continue
+			}
+
+		} else {
+			imgRef, err = imageapi.ParseDockerImageReference(status.Image.DockerImageReference)
+			if err != nil {
+				utilruntime.HandleError(fmt.Errorf(
+					"failed to parse isi.spec.repository.objectreference.from.name %q: %v",
+					status.Image.DockerImageReference, err))
+				continue
+			}
+		}
+
+		info := mkErrorInfo(imgRef.DockerClientDefaults().Registry, &status.Status)
+		cb(&info)
+	}
+}
+
+func mkErrorInfo(registry string, status *metav1.Status) metrics.ImportErrorInfo {
```

Comment:
calling this mkErrorInfo is a bit misleading since there may not be any error at all.

Reply:
Renamed.